docs/proposals/loadbalancing.md

```diff
@@ -0,0 +1,338 @@
+* **Claim**: A resource that represents a claim on an IP address.
+* **Load Balancer Controller**: a controller capable of fulfilling all Ingress paths for a given class of claims.
+
+__Example__:
+* Cluster is bootstrapped with kube-controller-manager --loadbalancer-controllers=(cloud=gce),haproxy
+* User creates svc1 and svc2
+* User creates claim1 to get a public ip.
+  GCEController:
+    - receives watch notification for `Class: cloud` object.
+    - allocates cloud resources for a new L7
+    - spins off a goroutine that watches services for that one claim and updates the L7
+* User creates claim2.
+  Haproxy Controller:
+    - receives watch notification for `Class: haproxy` object
+    - creates a haproxy pod
+    - haproxy pod watches services for that one claim and reconfigures haproxy
+* We end up with 2 ips:
+    - 172.12.14.111:80 serving foo and bar at L7 through a GCE L7
+    - 172.11.14.143:8080 switching between foobar and baz with different certs using sni at L4, through a pod in the cluster.
+
+```
+                Cloud                                  HaProxy pod
+                  ^                                         ^
+                  |                                         |
+            --------------------                     -------------------------
+           | GCE Controller pod |                   | Ha proxy Controller pod |
+            --------------------                     -------------------------
+                  |                                         |
+                  v                                         v
+            {Name: claim1,                          {Name: claim2,
+             Class: cloud,                           Class: bare-metal,
+             Ip: 172.11.14.111}                      Ip: 172.11.14.143:8080}
+                 /       \                              /        \
+             Path1      Path2                       Path3        Path4
+             type: l7   type: l7                    type: l4     type: l4
+             foo.com    bar.com                     foobar.com   baz.com
+             /a - svc1  /prod - svc2                cert: c1     cert: c2
+             /b - svc1                              svc4         svc5
+```
+
+#### Ingress Path
+
+This is an example ingress path. It encapsulates:
+* L7 proxying information (i.e map these urls to these backends)
+* Security configuration (i.e tlsMode, secret)
+* Metadata needed for more advanced loadbalancing (e.g session persistence)
+
+See [draft pr](https://github.com/kubernetes/kubernetes/pull/12825) for an actual description of the resource. See [alternatives](#alternatives) for why this deviates from existing implementations in OpenShift or GCE:
+
+```yaml
+apiVersion: v1
+kind: IngressPoint
+metadata:
+  name: l7ingress
+  type: l7
+spec:
+  host: example
+  tlsMode: Termination
+  secret: foosecret
+  pathMap:
+    "foo.example.com":
+    - url: "/foo/*"
```

Comment:
Api server will run through some basic validation for the path regex, but otherwise it's upto the loadbalancer controller to deal with it. if it wants to make /foo == /foo/* it can.

Renaming url to path sgtm